frontend/webapp/utils/ldap/forms.py

```diff
@@ -0,0 +1,122 @@
+from werkzeug.local import LocalProxy
+from wtforms import BooleanField
+from wtforms import PasswordField
+from wtforms import StringField
+from wtforms import SubmitField
+
+_datastore = LocalProxy(lambda: current_app.extensions["security"].datastore)
+
+
+class LDAPLoginForm(Form, NextFormMixin):
+    """Login form for LDAP users."""
+
+    email = StringField("User ID")
+    password = PasswordField(get_form_field_label("password"))
+    remember = BooleanField(get_form_field_label("remember_me"))
+    submit = SubmitField(get_form_field_label("login"))
+
+    def __init__(self, *args, **kwargs):
+        """Init new LDAP login form."""
+        super(LDAPLoginForm, self).__init__(*args, **kwargs)
+        self._args = args
+        self._kwargs = kwargs
+        if not self.next.data:
+            self.next.data = request.args.get("next", "")
+        self.remember.default = config_value("DEFAULT_REMEMBER_ME")
+
+    def validate(self):
+        """Validate LDAP logins against AD."""
+        if not super(LDAPLoginForm, self).validate():
+            return False
+
+        if self.email.data.strip() == "":
+            self.email.errors.append(get_message("USERID_NOT_PROVIDED")[0])
+            return False
+
+        if self.password.data.strip() == "":
+            self.password.errors.append(get_message("PASSWORD_NOT_PROVIDED")[0])
+            return False
+
+        try:
+            admin_user = _datastore.get_user(1)
+            if self.email.data == admin_user.username:
+                return self._try_local_auth()
+
+            # first we try authenticating against ldap
+            user_dn, ldap_data = _datastore.query_ldap_user(self.email.data)
+
+            if not _datastore.verify_password(user_dn, self.password.data):
+                self.password.errors.append(get_message("INVALID_PASSWORD")[0])
+                return False
+
+            ldap_email = ldap_data[config_value("LDAP_EMAIL_FIELDNAME")].value
+            password = encrypt_password(self.password.data)
+
+            if _datastore.find_user(email=ldap_email):
+                self.user = _datastore.get_user(ldap_email)
+                # note that this is being stored per user login
+                self.user.password = password
+            else:
+                self.user = _datastore.create_user(
+                    username=ldap_email,
+                    email=ldap_email,
+                    password=password,
+                    active=True,
+                )
+                # need to somehow decide what role they are
+                groups = config_value("LDAP_ADMIN_GROUPS")
+                if any(
+                    [
+                        group in ldap_data[config_value("LDAP_ADMIN_GROUPS_FIELDNAME")]
+                        for group in groups
+                    ]
+                ):
+                    role = _datastore.find_role("admin")
+                else:
+                    role = _datastore.find_role("user")
```

Comment:
so any user that is in the LDAP can connect to ARTEMIS? Maybe we want a more secure behavior where users and admins  are explicitly marked? Instead of "not-admin" ==> "user"

Reply:
you put a specific domain that is queried inside the configuration file.. not sure if we want to support multiple or not there

Reply:
so the domain will be e.g., "artemis", and only LDAP users that are in this group may login to ARTEMIS? (depending on the objectClass, they will then be deemed users or admins). Do I get this correctly? Is this supported in the current implementation or does it need a previous arbitration step? (to discard LDAP users that are not in the artemis group, either as users or admins)?

Reply:
The impl. will be based on a visibility domain, so this comment is auto-resolved for now.